python/cudf/cudf/core/index.py

```diff
@@ -2752,6 +2764,235 @@ def categories(self):
+    end : numeric , default None
+        Right bound for generating intervals.
+    periods : int, default None
+        Number of periods to generate
+    freq : numeric, default None
+        The length of each interval. Must be consistent
+        with the type of start and end
+    name : str, default None
+        Name of the resulting IntervalIndex.
+    closed : {"left", "right", "both", "neither"}, default "right"
+        Whether the intervals are closed on the left-side, right-side,
+        both or neither.
+
+    Returns
+    -------
+    IntervalIndex
+
+    Examples
+    --------
+    >>> import cudf
+    >>> import pandas as pd
+    >>> cudf.interval_range(start=0,end=5)
+    IntervalIndex([(0, 0], (1, 1], (2, 2], (3, 3], (4, 4], (5, 5]],
+    ...closed='right',dtype='interval')
+    >>> cudf.interval_range(start=0,end=10, freq=2,closed='left')
+    IntervalIndex([[0, 2), [2, 4), [4, 6), [6, 8), [8, 10)],
+    ...closed='left',dtype='interval')
+    >>> cudf.interval_range(start=0,end=10, periods=3,closed='left')
+    ...IntervalIndex([[0.0, 3.3333333333333335),
+            [3.3333333333333335, 6.666666666666667),
+            [6.666666666666667, 10.0)],
+            closed='left',
+            dtype='interval')
+    """
+    if freq and periods and start and end:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and "
+            "freq, exactly three must be specified"
+        )
+    args = [start, end, freq, periods]
+    *args, periods = [cudf.Scalar(x) if x is not None else None for x in args]
+    if any(
+        [
+            not is_numerical_dtype(x.dtype) if x is not None else False
+            for x in args
+        ]
+    ):
+        raise ValueError("start, end, freq must be numeric values.")
+    common_dtype = find_common_type([x.dtype for x in args if x])
+    start, end, freq = args
+
+    if periods and not freq:
+        # if statement for mypy to pass
+        if end is not None and start is not None:
+            # determine if periods are float or integer
+            periods = periods._as_host_type("int64")
+            # divmod only supported on host side scalars
+            quotient, remainder = divmod((end - start).value, periods.value)
+            if remainder:
+                freq_step = cudf.Scalar((end - start) / periods)
+            else:
+                freq_step = cudf.Scalar(quotient)
+            common_dtype = find_common_type([common_dtype, freq_step.dtype])
+            if start.dtype != freq_step.dtype:
+                start = start._as_host_type(freq_step.dtype)
+            bin_edges = sequence(
+                size=periods + 1,
+                init=start.device_value,
+                step=freq_step.device_value,
+            )
+            left_col = bin_edges[:-1].astype(common_dtype)
+            right_col = bin_edges[1:].astype(common_dtype)
+    elif freq and periods:
+        if end:
+            start = end - (freq * periods)
```

Comment:
I think we need to either cast `periods` to an integer or throw if it's not an integer before this otherwise we can get incorrect results

Reply:
Pandas handles it being a float by casting to int it seems

Reply:
Performed a coerce for periods to `int64` at top level.